firmware/lib/Gantry/Axis.cpp

```diff
@@ -1,311 +1,569 @@
+ * so the correct TC?_Handler functions can be defined
+ */
+
+#ifndef AXIS_X_STEP_TC_IRQ
+#error "AXIS_X_STEP_TC_IRQ must be defined"
+#endif //AXIS_X_VEL_TC_IRQ
+
+#ifndef AXIS_Y_STEP_TC_IRQ
+#error "AXIS_Y_STEP_TC_IRQ must be defined"
+#endif //AXIS_Y_VEL_TC_IRQ
+
+/*****************************************************************************/
+/*                                 TYPEDEFS                                  */
+/*****************************************************************************/
+
+/**
+ * @struct AxisInterrupts
+ * 
+ * @brief Specifies all of the internal interrupts and ISRs used by the axis
+ */
+typedef struct {
+    Tc * const timer;                 //!< Timer counter
+    const uint32_t channel_accel;     //!< Timer counter channel for the acceleration interrupt
+    const IRQn_Type irq_accel;        //!< IRQ number for the acceleration interrupt
+
+    void (*const isr_encoder)(void);  //!< ISR for the encoder interrupt
+    void (*const isr_ls_home)(void);  //!< ISR for the home limit switch interrupt
+    void (*const isr_ls_far)(void);   //!< ISR for the far limit switch interrupt
+} AxisInterrupts;
+
+/**
+ * @struct Axis
+ * 
+ * @brief Top-level struct for encapsulating all elements of an axis
+ */
+typedef struct {
+    AxisIO io;                         //!< I/O pins and peripherals
+    const AxisInterrupts interrupts;   //!< Interrupt configurations
+    AxisMotion motion;                 //!< Profile for the current motion
+    AxisState state;                   //!< Current state of the axis
+} Axis;
+
+/*****************************************************************************/
+/*                             AXIS DECLARATIONS                             */
+/*****************************************************************************/
+
+/* ******************************** X AXIS ********************************* */
 
-// X AXIS
-#define ACCEL_X                 8000
-#define VEL_MIN_X               500
-#define VEL_MAX_X               25000
-#define STEP_PIN_X              5 // need to know the register for fast operation pin 5 - PC25 
-#define DIR_PIN_X               6 // need to know the register for fast operation pin 6 - PC24
-#define ENCODER_PIN_A_X         7
-#define ENCODER_PIN_B_X         8
-#define LIMIT_SW_HOME_PIN_X     9
-#define LIMIT_SW_FAR_PIN_X      10
-
-// Y AXIS
-#define ACCEL_Y                 8000
-#define VEL_MIN_Y               500
-#define VEL_MAX_Y               25000
-#define STEP_PIN_Y              22 // need to know the register for fast operation pin 22 - PB26 
-#define DIR_PIN_Y               23 // need to know the register for fast operation pin 23 - PA14
-#define ENCODER_PIN_A_Y         24
-#define ENCODER_PIN_B_Y         25
-#define LIMIT_SW_HOME_PIN_Y     26
-#define LIMIT_SW_FAR_PIN_Y      27
-
-// Interrupt Declarations
 void isr_encoder_x();
-void isr_limit_switch_x();
+void isr_ls_home_x();
+void isr_ls_far_x();
+
+static Axis axis_x = {
+    .io = {},
```

Comment:
Didn't know you could do this dot thing, pretty sweet!

Reply:
Yup, great way to initialize structs!